pkg/controller/daemon/update.go

```diff
@@ -172,6 +193,27 @@ func (dsc *DaemonSetsController) rollingUpdate(ctx context.Context, ds *apps.Dae
 				continue
 			default:
 				logger.V(5).Info("DaemonSet pod on node is out of date, this is a surge candidate", "daemonset", klog.KObj(ds), "pod", klog.KObj(oldPod), "node", klog.KRef("", nodeName))
+
+				node, err := dsc.nodeLister.Get(nodeName)
+				if err != nil {
+					return fmt.Errorf("couldn't get node for nodeName %q: %v", nodeName, err)
+				}
+
+				shouldRun, shouldContinueRunning := NodeShouldRunDaemonPod(node, ds)
+
+				// daemonset should not run pod on the node
+				if !shouldRun {
+					logger.V(5).Info("Exclude the node from daemon pod updating", "daemonset", klog.KObj(ds), "pod", klog.KObj(oldPod), "node", klog.KRef("", nodeName))
+
+					// daemonset should continue running pod on the node
+					if shouldContinueRunning {
```

Comment:
The above comment applies even here for the available pods but it is more complicated.

First we should not keep the revision A pods running, or we will clutter nodes with our old pods. Eg just try to move a DS from a set of nodes Y to a set of nodes Z and you will have a DS tracking 3 pods even though 6 are running.

Secondly, due to our validation, maxUnavailable must be set to 0 when surge is enabled. But the problem is that is if you change scheduling constraints from set of nodes Y to a set of nodes Z and there is no intersection for these nodes, the application would not have a single available pod at a time and thus possibly disrupting some applications.

Thus we probably should delete these pods after all of the new pods have become available to preserve the availability (one of the ways would be to allow maxaUnavailable feature, but that is more complicated and probably deserves a KEP)

We could internally implement maxUnavailable=1, but that would still not work for eg a two node cluster where the daemonset would be active only on a single node.

IMO, best way to do it is decrementally delete the pods when "numberAvailable > desiredNumberScheduled" (in `rollingUpdate` function), ie we delete a single pod from an old set Y, when a new pod becomes available from a set Z.

Reply:
@atiratree 
Thank you for your advice!

> IMO, best way to do it is decrementally delete the pods when "numberAvailable > desiredNumberScheduled" (in rollingUpdate function), ie we delete a single pod from an old set Y, when a new pod becomes available from a set Z.

I'll fix as you said.
However numberAvailable and desiredNumberScheduled seems to be calculated after evaluation of shouldRun.
If there are daemon pods on the node not matching the scheduling constraints, numberAvailable will not be incremented.

https://github.com/kubernetes/kubernetes/blob/97c7dbcd227ec253bb277a91575706fe30f519dc/pkg/controller/daemon/daemon_controller.go#L1174-L1190

So I think we need to fix the logic of updating DaemonSet status, right?

Reply:
@mochizuki875  I think we can count all the values we need in situ. Also take a look if we can use `updatedDesiredNodeCounts` and maybe return additional values? Or if it makes more sense to count it in `rollingUpdate`. I mentioned `numberAvailable` and `desiredNumberScheduled` just for an illustration; and they do not have to match the status. There are more ways how to implement this.

Nevertheless, I would not change the status counting or updating logic, unless there is a good reason for it.

Reply:
@atiratree 
Thank you for your comment.
I've fixed that.
Could you please check again?